### PR TITLE
NamedPipeClientStream Connect with SafePipeHandle

### LINKED ISSUE #32760: NamedPipeClientStream Connect can use uninitialized _normalizedPipePath
When adding nullable annotations for System.Net.Sockets (#32675), it was discovered that there is a scenario when using `NamedPipeClientStream` can use a null string:

When calling `NamedPipeClientStream.Connect` on Unix, it will try using the `_normalizedPipePath` string:

https://github.com/dotnet/runtime/blob/032e8dbd0ca5aee045a41ff78e3c0244d3ec3b51/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Unix.cs#L21-L31

However, if the `NamedPipeClientStream` object was created with the constructor that takes an existing `SafePipeHandle` and `isConnected: false`, then `_normalizedPipePath` will be `null`.

https://github.com/dotnet/runtime/blob/032e8dbd0ca5aee045a41ff78e3c0244d3ec3b51/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.cs#L101-L103

Doing some searching, I'm honestly not certain how this scenario (existing `SafePipeHandle` and `isConnected: false`) can ever work successfully. It will need a `path` to be able to connect to, which it wasn't given. So at best we may be able to throw a better exception than the `ArgumentNullException` we get:

```
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'path')
   at System.Net.Sockets.UnixDomainSocketEndPoint..ctor(String path)
   at System.IO.Pipes.NamedPipeClientStream.TryConnect(Int32 timeout, CancellationToken cancellationToken)
   at System.IO.Pipes.NamedPipeClientStream.ConnectInternal(Int32 timeout, CancellationToken cancellationToken, Int32 startTime)
   at System.IO.Pipes.NamedPipeClientStream.Connect(Int32 timeout)
   at System.IO.Pipes.NamedPipeClientStream.Connect()
```

/cc @stephentoub

### PR DESCRIPTION
**Pull request by koenigst**
* Checked that isConnected in the NamedPipeClientStream constructor is always true.
* Created a new constructor without the argument.
* Marked the old constructor as obsolete.
* Corrected the nullability in the interop declarations.

I am aware that I did slightly more than in the original issue. Therefore I created a base commit and two additional commits that can be easily removed from this PR if desired.

Fixes #32760

### CODE REVIEW

File: src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.cs [hunk 1]
```diff
@@ -89,7 +89,18 @@ public NamedPipeClientStream(string serverName, string pipeName, PipeDirection d
...
+            : this(direction, isAsync, safePipeHandle)
+        {
+            if (!isConnected)
+            {
+                throw new ArgumentOutOfRangeException(nameof(isConnected));
+            }
+        }
+
+        // Create a NamedPipeClientStream from an existing server pipe handle.
+        public NamedPipeClientStream(PipeDirection direction, bool isAsync, SafePipeHandle safePipeHandle)
```
Comment (huoyaoyuan) [inline src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.cs:103]: This would need API review, which isn't present in linked issue.

**Comment by stephentoub** [comment]
Thanks for trying to help out! New surface area needs to first be reviewed and approved, following:
https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md
Either a new issue should be opened proposing the new surface area, or the existing issue should be updated into a proposal. Either way, though, we can't accept this PR until such time that a new member has been agreed upon, so I'm going to close this. If/when an API is approved, you're of course welcome to re-open. Thanks!

**Comment by koenigst** [comment]
@stephentoub I'm sorry for the inconvenience. I created a new PR (#110175) without the changes to the public API.

**Comment by stephentoub** [comment]
No worries. Thanks!

**Pull request closed without merging**